NWNXLib/Services/Commands/Commands.cpp

```diff
@@ -0,0 +1,128 @@
+#include "Services/Commands/Commands.hpp"
+#include <stdio.h>
+#include <dlfcn.h>
+
+namespace NWNXLib {
+namespace Services {
+
+static Commands *g_commands;
+
+Commands::Commands()
+{
+    g_commands = this;
+    RegisterCommand("nwnx_commands_test", [](std::string& s){ LOG_NOTICE("Selftest command. Args: '%s'", s.c_str()); });
+}
+
+Commands::~Commands()
+{
+    g_commands = nullptr;
+
+}
+
+bool Commands::RegisterCommand(const std::string& cmd, CommandFunc func)
+{
+    auto it = m_commandMap.find(cmd);
+    if (it == m_commandMap.end())
+    {
+        m_commandMap[cmd] = func;
+        LOG_INFO("Registering command '%s'", cmd.c_str());
+        return true;
+    }
+    else
+    {
+        LOG_WARNING("Command '%s' already registered", cmd.c_str());
+        return false;
+    }
+}
+
+void Commands::UnregisterCommand(const std::string& cmd)
+{
+    LOG_INFO("Unregistering command '%s'", cmd.c_str());
```

Comment:
I guess there is a minor issue here - plugins can unregister each other's commands. It's minor and I don't care much about it. The reason the proxy exists is to avoid this - each plugin has its own instance of the proxy which stores instance-specific info (e.g. which commands it has already registered) to ensure it can only touch what it owns. Not saying you need to change it for me to merge this PR, just stating it here for consideration.

Reply:
You're right. I'll fix that before we should merge this. There's also the issue of plugin hotplug/unload - we want to unregister all of its commands, so that next time it loads it'll be available. Hold the PR for now.